Puneet/feature/edit feature attributes

puneet-pdx opened the pull request:
UI and functionality for Feature editing

<img width="350" alt="phoneProfile" src="https://user-images.githubusercontent.com/5582469/94496130-5f4dcd80-01a8-11eb-8389-73eaef6c4f77.png">

<img width="350" alt="phoneProfile" src="https://user-images.githubusercontent.com/5582469/94496126-5ceb7380-01a8-11eb-866f-ac66bf160351.png">

<img width="350" alt="phoneProfile" src="https://user-images.githubusercontent.com/5582469/94496040-2b72a800-01a8-11eb-82d1-a7d6e97a1074.png">

<img width="350" alt="phoneProfile" src="https://user-images.githubusercontent.com/5582469/97389476-98f12180-1897-11eb-9f24-afccbde6c3de.png">

puneet-pdx committed:
WIP

puneet-pdx committed:
WIP

puneet-pdx committed:
WIP

puneet-pdx committed:
Merge branch 'v.next' into puneet/feature/edit_feature_attributes

puneet-pdx committed:
WIP

puneet-pdx committed:
add doc

sorenoid commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/popup/PopupView.kt:
```diff
@@ -17,22 +17,39 @@
 import kotlinx.android.synthetic.main.layout_popupview.view.*
+import java.text.ParseException
+import java.text.SimpleDateFormat
+import java.util.*
```
```suggestion
import java.util.Calendar
import java.util.Locale
```

sorenoid commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/popup/PopupView.kt:
```diff
@@ -17,22 +17,39 @@
 import androidx.recyclerview.widget.RecyclerView
+import com.esri.arcgisruntime.ArcGISRuntimeException
+import com.esri.arcgisruntime.concurrent.ListenableFuture
+import com.esri.arcgisruntime.data.*
```
```suggestion
import com.esri.arcgisruntime.data.Feature
import com.esri.arcgisruntime.data.FeatureEditResult
import com.esri.arcgisruntime.data.FeatureTable
import com.esri.arcgisruntime.data.Field
import com.esri.arcgisruntime.data.ServiceFeatureTable
```

sorenoid commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/popup/PopupView.kt:
```diff
@@ -71,28 +123,178 @@ class PopupView : FrameLayout {
+                break
+            }
+        }
+        return errorEncountered
```
this doesn't change anything, but seems like a good spot for some functional programming

```suggestion
       featureEditResults
            .filter { result ->
                result.hasCompletedWithErrors()
            }
            .first { return true }

        return false
```

puneet-pdx replied:
updated to use any()

sorenoid commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/popup/PopupView.kt:
```diff
@@ -17,22 +17,39 @@
 import androidx.recyclerview.widget.RecyclerView
+import com.esri.arcgisruntime.ArcGISRuntimeException
+import com.esri.arcgisruntime.concurrent.ListenableFuture
+import com.esri.arcgisruntime.data.*
```
Should probably flesh out the data.* imports.

puneet-pdx replied:
updated

sorenoid commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/popup/PopupView.kt:
```diff
@@ -17,22 +17,39 @@
 import kotlinx.android.synthetic.main.layout_popupview.view.*
+import java.text.ParseException
+import java.text.SimpleDateFormat
+import java.util.*
```
Should probably flesh out the java.util.* imports

puneet-pdx replied:
updated

sorenoid commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/popup/PopupView.kt:
```diff
@@ -71,28 +123,178 @@ class PopupView : FrameLayout {
+                break
+            }
+        }
+        return errorEncountered
```
could use `any()` here, though the results are the same.

```suggestion
        return featureEditResults.any { result -> result.hasCompletedWithErrors() }
```

puneet-pdx replied:
updated

sorenoid commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/popup/PopupView.kt:
```diff
@@ -132,22 +338,97 @@ class PopupView : FrameLayout {
+         * and if the popup attribute is editable.
+         */
+        fun isEditable(popupAttribute: PopupAttribute): Boolean {
+            return editMode && popupAttribute.isEditable
```
This is just a style question, is there a kotlin preferred style for member variable names? I was expecting something like `mEditMode` when I was looking for the reasoning of making this class `inner` and I noticed that none of the members of PopupView use this style.

puneet-pdx replied:
the `mEditMode` is java style. For Kotlin see here
https://kotlinlang.org/docs/reference/coding-conventions.html#naming-rules

sorenoid commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/popup/PopupView.kt:
```diff
@@ -71,28 +123,178 @@ class PopupView : FrameLayout {
+            } else {
+                manager.updateValue(value.toString().toDouble(), field)
+            }
+            Field.Type.DATE -> {
```
you mentioned in slack this PR doesn't handle Calendar input. Is this case intentional?

puneet-pdx replied:
This here just updates the local geodatabase with the new values that the user enters. We are still not popping a calendar to help user enter dates via a calendar prompt...

sorenoid commented on app/src/main/res/layout/fragment_popup_attribute_list.xml:
```diff
@@ -59,18 +61,70 @@
 
+            <ImageView
+                    android:id="@+id/edit"
+                    android:layout_width="30dp"
+                    android:layout_height="34dp"
```
The sizes all seem to be hardcode here and in the related row xml. Have you tested this out on different form factors, or something like that?

puneet-pdx replied:
I have tested this out on a phone and a tablet. The UI still needs work though and I will be addressing it in the next issue.

sorenoid reviewed (changes requested):
Hi @puneet-pdx this looks good, and I love the app. I have just a few minor points and a few questions too.

puneet-pdx committed:
clean up wildcard imports

puneet-pdx committed:
update error checking logic

puneet-pdx commented:
@sorenoid I have addressed your comments.

eri9000 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/popup/PopupView.kt:
```diff
@@ -71,28 +128,164 @@ class PopupView : FrameLayout {
+     * Shows the message in a Snackbar.
+     */
+    private fun showSnackbarMessage(message: String) {
+        val snackbar: Snackbar = Snackbar.make(this, message, Snackbar.LENGTH_LONG)
+        snackbar.show()
```
This can be shortened?
```suggestion
        Snackbar.make(this, message, Snackbar.LENGTH_LONG).show()
```

puneet-pdx replied:
updated

eri9000 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/popup/PopupView.kt:
```diff
@@ -71,28 +128,164 @@ class PopupView : FrameLayout {
+        var error: ArcGISRuntimeException? = null
+
+        when (manager.getFieldType(field)) {
+            Field.Type.SHORT -> error = if (value == null || value.trim { it <= ' ' }.isEmpty()) {
```
I think this might work better, also applies to all other if statement below 

https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.text/is-null-or-blank.html
```suggestion
            Field.Type.SHORT -> error = if (value. isNullOrBlank()){
```

puneet-pdx replied:
I have removed this code. No longer applicable.

eri9000 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/popup/PopupView.kt:
```diff
@@ -71,28 +128,164 @@ class PopupView : FrameLayout {
+            Field.Type.SHORT -> error = if (value == null || value.trim { it <= ' ' }.isEmpty()) {
+                manager.updateValue(null, field)
+            } else {
+                manager.updateValue(value.toString().toShort(), field)
```
I don't think there's need to call `toString()` again since value is already a String, applies to all others below 
```suggestion
                manager.updateValue(value.toShort(), field)
```

puneet-pdx replied:
I have removed this code. No longer applicable.

eri9000 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/popup/PopupView.kt:
```diff
@@ -71,28 +128,164 @@ class PopupView : FrameLayout {
+                            "M/dd/yy hh:m a",
+                            Locale.getDefault()
+                        ).parse(value)
+                        calendar.timeInMillis = date.time
```
The compiler gives a warning here `Unsafe use of a nullable receiver of type Date?`
But I think this can't be null, so we can just do: 
```suggestion
                        calendar.timeInMillis = date!!.time
```

puneet-pdx replied:
I have removed this code. No longer applicable.

eri9000 reviewed (commented):
@puneet-pdx 
Code looks good, Left some comments.

puneet-pdx committed:
Removed usage of PopupAttribute, relying on PopupManager for implementation

puneet-pdx committed:
removed variable assignment

puneet-pdx commented:
@eri9600 I have addressed your comments(sort of) .  I have changed the implementation to not use PopupAttribute data type and rely on PopupManager to maintain/do the edits.

scottlehto commented:
Awesome. That’s great. I will look into it for sure in the next few weeks.

 

Thanks,

Scott

 

 

From: Puneet Prakash <notifications@github.com> 
Sent: Monday, October 5, 2020 8:00 PM
To: Esri/data-collection-android <data-collection-android@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Subject: Re: [Esri/data-collection-android] Puneet/feature/edit feature attributes (#9)

 

@eri9600 <https://github.com/eri9600>  I have addressed your comments(sort of) . I have changed the implementation to not use PopupAttribute data type and rely on PopupManager to maintain/do the edits.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub <https://github.com/Esri/data-collection-android/pull/9#issuecomment-703950987> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADZRMKQ4DATN3VARTNZK5VLSJJMX7ANCNFSM4RU6JGNA> .  <https://github.com/notifications/beacon/ADZRMKU3BB2BM66F4ULWPWDSJJMX7A5CNFSM4RU6JGNKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOFH2XBCY.gif>

eri9000 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/popup/PopupView.kt:
```diff
@@ -44,9 +60,36 @@ class PopupView : FrameLayout {
+        }
+
+    var saveEdit: Boolean = false
+        set(value) {field = value
```
This could go in a new line 
```suggestion
        set(value) {
        field = value
```

puneet-pdx replied:
No longer relevant

eri9000 reviewed (approved):
@puneet-pdx 
Code looks good.
I tested on my device and it also seems to work.
One small bug, when i try to edit a Popup and and I click on a TextView, the keyboard appears but then I click the back top left button or the X button and press back to go back to the main Activity, the keyboard stays on the screen. 
Would you like me to create an issue for this?

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/viewmodels/IdentifyResultViewModel.kt:
```diff
@@ -83,6 +92,30 @@ class IdentifyResultViewModel : ViewModel() {
+     * Called when user taps on the identified popup attribute list header in the expanded bottom sheet
+     * Kicks off IdentifyResultFragment
+     */
+    fun showIdentifiedPopupAttribute() {
```
Should be plural? `showIdentifiedPopupAttributes()`?

puneet-pdx replied:
updated

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/viewmodels/IdentifyResultViewModel.kt:
```diff
@@ -83,6 +92,30 @@ class IdentifyResultViewModel : ViewModel() {
+     * Kicks off IdentifyResultFragment
+     */
+    fun showIdentifiedPopupAttribute() {
+        _showIdentifiedPopupAttributeEvent.raiseEvent()
```
as above, plural?

puneet-pdx replied:
updated

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/popup/PopupView.kt:
```diff
@@ -44,9 +60,36 @@ class PopupView : FrameLayout {
+            if (field) {
+                popupAttributeListAdapter.submitList(popupManager.editableFields)
+                popupManager.startEditing()
+            } else if (popup != null) {
```
Is this null check required here? Wouldn't we want to show display fields regardless of popup being null?

puneet-pdx replied:
no longer relevant

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/popup/PopupView.kt:
```diff
@@ -44,9 +60,36 @@ class PopupView : FrameLayout {
+                popupManager.startEditing()
+            } else if (popup != null) {
+                popupAttributeListAdapter.submitList(popupManager.displayedFields)
+                if (!saveEdit) {
```
Shouldn't switching the edit mode to false always cancel editing? In other words, if edits should be saved, this should always happen before the edit mode is switched back to false.

puneet-pdx replied:
I have refactored to have the cancel editing logic separate from finish editing and saving popup. This logic is no longer there.

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/popup/PopupView.kt:
```diff
@@ -44,9 +60,36 @@ class PopupView : FrameLayout {
     var popup: Popup? = null
```
Should `popup` be nullable? I assume a PopupView always has to be associated with a non-null `Popup`

puneet-pdx replied:
I had a custom setter for popup which required popup to be initialized. This logic is no longer in place.

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/popup/PopupView.kt:
```diff
@@ -71,35 +114,73 @@ class PopupView : FrameLayout {
      */
-    data class PopupAttribute(val field: String, val value: String)
+    private fun savePopup() {
+        val editingErrorFuture: ListenableFuture<ArcGISRuntimeException> = popupManager.finishEditingAsync()
```
naming suggestion: `val finishEditingFuture:`

puneet-pdx replied:
updated

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/popup/PopupView.kt:
```diff
@@ -71,35 +114,73 @@ class PopupView : FrameLayout {
+                        ie
+                    )
+                )
+            } catch (ee: ExecutionException) {
```
No need for two catch blocks. Just catch `Exception` and log message. I wonder if this needs more user feedback other than just a log.

puneet-pdx replied:
I have removed two catch blocks. 
In case of no network on the client side. An exception is thrown when the app is not able to apply the edits. I am showing the message of the exception in a dialog box to the user.

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/popup/PopupView.kt:
```diff
@@ -71,35 +114,73 @@ class PopupView : FrameLayout {
+                                    ie
+                                )
+                            )
+                        } catch (ee: ExecutionException) {
```
No need for two catch blocks

puneet-pdx replied:
updated

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/popup/PopupView.kt:
```diff
@@ -71,35 +114,73 @@ class PopupView : FrameLayout {
+                            val featureEditResults: List<FeatureEditResult> = applyEditsFuture.get()
+                            // Check for errors in FeatureEditResults
+                            if (featureEditResults.any { result -> result.hasCompletedWithErrors() }) {
+                                showSnackbarMessage("There was a problem saving the feature")
```
Do we always expect a ServiceFeatureTable to be edited? If not, then we also need to show this for other types of feature tabes.

puneet-pdx replied:
At this time we only support ServiceFeatureTable. The offline editing workflows will come later.

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/popup/PopupView.kt:
```diff
@@ -71,35 +114,73 @@ class PopupView : FrameLayout {
+                            if (featureEditResults.any { result -> result.hasCompletedWithErrors() }) {
+                                showSnackbarMessage("There was a problem saving the feature")
+                            } else {
+                                showSnackbarMessage("Feature saved successfully")
```
As above, need to show also for other types of feature tabes?

puneet-pdx replied:
No, just for ServiceFeatureTable at this time

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/popup/PopupView.kt:
```diff
@@ -112,42 +193,127 @@ class PopupView : FrameLayout {
+         * Toggles the view for popup field value from edittext to textview and vice-versa, given the
+         * edit mode of the popupView.
+         */
+        fun setViewsBehavior(popupField: PopupField) {
```
Would `updateViewMode` be more descriptive?

puneet-pdx replied:
updated.

puneet-pdx commented:
> @puneet-pdx
> Code looks good.
> I tested on my device and it also seems to work.
> One small bug, when i try to edit a Popup and and I click on a TextView, the keyboard appears but then I click the back top left button or the X button and press back to go back to the main Activity, the keyboard stays on the screen.
> Would you like me to create an issue for this?

@eri9600 right now we don't control the dismissing of the keyboard. It is up to the user. The user can minimize the keyboard using the back button on their device.

puneet-pdx committed:
Refactored PopupView to have an associated PopupViewModel

puneet-pdx committed:
added missing logic to start the editing on popupManager again after failed applyeditsAsync

puneet-pdx committed:
moved cancelEditing() from popupView

sorenoid commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/fragments/PopupAttributesListFragment.kt:
```diff
@@ -43,11 +50,73 @@ class PopupAttributesListFragment : Fragment() {
+        dialogBuilder.setMessage(message)
+            .setCancelable(false)
+            // positive button text and action
+            .setPositiveButton("OK") { dialog, id ->
```
nitpicky suggestion: you've got strings for OK and Cancel, might as well use them.

sorenoid replied:
I just realized they are differently cased, but they are only used for accessibility controls in layout, and so could be made to have the same capitalization.

puneet-pdx replied:
good catch. Updated to use the resource strings.

sorenoid commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/popup/PopupView.kt:
```diff
@@ -112,42 +117,93 @@ class PopupView : FrameLayout {
+            newItem: PopupField
         ): Boolean {
-            return oldItem.field == newItem.field && oldItem.value == newItem.value
+            return oldItem.fieldName == newItem.fieldName
```
I think you should use String's equals method here.

```suggestion
            return oldItem.fieldName.equals(newItem.fieldName)
```

puneet-pdx replied:
The "==" operator in Kotlin works different than Java, it checks the contents for comparison. 

https://kotlinlang.org/docs/reference/equality.html

sorenoid reviewed (approved):
@puneet-pdx These changes are good. I would consider changing the name of PopupViewModel to PopupAttributeListViewModel, as it works on behalf of the PopupView and the edit controls at the top of the fragment, but it isn't really a technical issue. I didn't find anything to take issue with in the code, although I have some suggestions you will see below.

I tested the app a bit and found no issues. I have some trees picked out to visit 👍

puneet-pdx commented:
> @puneet-pdx These changes are good. I would consider changing the name of PopupViewModel to PopupAttributeListViewModel, as it works on behalf of the PopupView and the edit controls at the top of the fragment, but it isn't really a technical issue. I didn't find anything to take issue with in the code, although I have some suggestions you will see below.
> 
> I tested the app a bit and found no issues. I have some trees picked out to visit 👍

@sorenoid The intention is to have the PopupViewModel as a toolkit component that will be used by the Data Collection app. That is why I kept the name generic (PopupViewModel) to be tied to the PopupView and not specific to the the usage in the app.

puneet-pdx committed:
addressed code review comments
reverted change to keep app in edit mode on sync failure

eri9000 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/viewmodels/IdentifyResultViewModel.kt:
```diff
@@ -23,34 +23,29 @@ import com.esri.arcgisruntime.data.Feature
     private val _showPopupAttributeListEvent = MutableLiveData<Event<Unit>>()
     val showPopupAttributeListEvent: LiveData<Event<Unit>> = _showPopupAttributeListEvent
 
+
```
extra line not needed

eri9000 reviewed (approved):
@puneet-pdx 
Code changes look good
Just left a comment about spacing.

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/viewmodels/PopupViewModel.kt:
```diff
@@ -0,0 +1,145 @@
+
+/**
+ * The view model for PopupView, that is responsible for maintaining the PopupView
+ * editMode state for orientation changes and for performing async operations.
```
I think this doc could do with some refinement. Something like:

The view model that represents a Popup. It supports:
- viewing a GeoElement's attributes
- editing a GeoElement's attributes as well as saving

A PopupViewModel can be bound to a PopupView for visualisation of the Popup GeoElement's attributes and editing experience.

puneet-pdx replied:
updated

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/viewmodels/PopupViewModel.kt:
```diff
@@ -0,0 +1,145 @@
+class PopupViewModel(application: Application) : AndroidViewModel(application) {
+
+    private val _identifiedPopup = MutableLiveData<Popup>()
+    val identifiedPopup: LiveData<Popup> = _identifiedPopup
```
The term "identify" should be avoided in this context. A PopupViewModel should only have knowledge of a "popup". Where this Popup instance is coming from, is irrelevant in this context.

gunt0001 replied:
Each of the public properties should probably have some doc. Will only mention it here..

puneet-pdx replied:
updated

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/viewmodels/PopupViewModel.kt:
```diff
@@ -0,0 +1,145 @@
+     *
+     * @param identifyLayerResult
+     */
+    fun processIdentifyLayerResult(
```
avoid term "identify", also in doc above

puneet-pdx replied:
updated

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/viewmodels/PopupViewModel.kt:
```diff
@@ -0,0 +1,145 @@
+    /**
+     * Resets the result popup of the identify operation
+     */
+    fun resetIdentifiedPopup() {
```
Seems more like a "clear" rather than a "reset"

puneet-pdx replied:
updated

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/viewmodels/PopupViewModel.kt:
```diff
@@ -0,0 +1,145 @@
+    /**
+     * Raises ConfirmCancelPopupEditingEvent that can be observed and used for
+     * prompting user with confirmation dialog to make sure the user wants to cancel edits.
+     * To be followed by cancelEditing() if the user respond is positive.
```
response

puneet-pdx replied:
updated

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/viewmodels/PopupViewModel.kt:
```diff
@@ -0,0 +1,145 @@
+     * prompting user with confirmation dialog to make sure the user wants to cancel edits.
+     * To be followed by cancelEditing() if the user respond is positive.
+     */
+    fun confirmAndCancelEditing() {
```
shorter: `confirmCancelEditing` ?

puneet-pdx replied:
updated to `confirmCancelEditing`

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/viewmodels/PopupViewModel.kt:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    /**
+     * Saves the popup by applying changes to the feature service.
```
Should it say?

Saves Popup edits by applying changes to the feature service associated with a Popup's feature.

puneet-pdx replied:
updated

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/viewmodels/PopupViewModel.kt:
```diff
@@ -0,0 +1,145 @@
+    /**
+     * Saves the popup by applying changes to the feature service.
+     */
+    fun savePopup() {
```
Maybe we should rename this to `savePopupEdits`?

puneet-pdx replied:
updated

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/viewmodels/PopupViewModel.kt:
```diff
@@ -0,0 +1,145 @@
+     */
+    fun savePopup() {
+        // show the Progress bar informing user that save operation is in progress
+        _toggleSavingPopupProgressBarEvent.raiseEvent(true)
```
Would `showSavingProgressEvent` be more descriptive?

puneet-pdx replied:
updated

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/viewmodels/PopupViewModel.kt:
```diff
@@ -0,0 +1,145 @@
+                    }
+                } catch (exception: Exception) {
+                    exception.message?.let { exceptionMessage ->
+                        _showSavePopupErrorEvent.raiseEvent(exceptionMessage)
```
Do you need to hide the progress bar here?

puneet-pdx replied:
yes, updated

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/viewmodels/PopupViewModel.kt:
```diff
@@ -0,0 +1,145 @@
+                    finishEditingFuture.get()
+
+                    // The edits were applied successfully to the local geodatabase,
+                    // push those changes to the ServiceFeatureTable by calling applyEditsAsync()
```
That comment is not correct, it should say something like:
// push those changes to the feature service by calling applyEditsAsync()

puneet-pdx replied:
updated

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/fragments/DataCollectionFragment.kt:
```diff
@@ -61,6 +62,8 @@ class DataCollectionFragment : Fragment() {
     // use by activityViewModels()
     private val mapViewModel: MapViewModel by activityViewModels()
 
+    private val popupViewModel: PopupViewModel by activityViewModels()
```
I'm not sure DataCollectionFragment needs a reference to the PopupViewModel. I think this could all be handled by IdentifyResultViewModel.

puneet-pdx replied:
There are two calls being made to functions in the PopupViewModel in DataCollectionFragment. One is to set the Popup in the PopupViewModel and second one is to cancel editing when user presses back. 

The second use is probably going to go away to mimic what iOS has, which is to not let user back out of editing by using the back button. But we still need to set the Popup on the PopupViewModel...

gunt0001 replied:
Setting the Popup in the PopupViewModel can be done in `IdentifyResultViewModel.processIdentifyLayerResult()` Essentially the IdentifyResultViewModel can fully encapsulate the PopupViewModel. It's about separation of concerns, in other words, should the DataCollectionFragment be concerned about the PopupViewModel? Since DataCollectionFragment is our main fragment, I would argue it shouldn't, otherwise this fragment will grow ever larger and become unmaintainable eventually.

puneet-pdx replied:
updated

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/fragments/PopupAttributesListFragment.kt:
```diff
@@ -43,11 +50,69 @@ class PopupAttributesListFragment : Fragment() {
         binding.lifecycleOwner = this
 
+        popupViewModel.isPopupInEditMode.observe(viewLifecycleOwner, Observer {
+            popupView.setEditMode(it)
```
Can't we do this via binding in the layout?

puneet-pdx replied:
Yes, we can. Updated

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/fragments/PopupAttributesListFragment.kt:
```diff
@@ -20,18 +20,25 @@ import android.os.Bundle
 /**
  * Responsible for displaying PopupAttribute list.
  */
 class PopupAttributesListFragment : Fragment() {
```
I wonder if we should rename this Fragment, since we don't have a corresponding "PopupAttributesListViewModel". Would it make more sense to rename this to "IdentifyResultFragment"?

puneet-pdx replied:
We already have `IdentifyResultFragment` . `PopupAttributesListFragment` mostly uses the PopupViewModel now . Before that `PopupAttributesListFragment`and `IdentifyResultFragment` shared the IdentifyResultViewModel as they both work off of the same Popup.

gunt0001 replied:
Good point. How about renaming it to `PopupFragment`? Then we would have the three related components - PopupViewModel, PopupView and PopupFragment - consistently named.

puneet-pdx replied:
updated

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/fragments/PopupAttributesListFragment.kt:
```diff
@@ -43,11 +50,69 @@ class PopupAttributesListFragment : Fragment() {
+    }
+
+    /**
+     * Shows error message to the use in a dialog.
```
user

puneet-pdx replied:
updated

gunt0001 replied:
I don't see the update

puneet-pdx replied:
Some how I missed to push it to remote. Updated now.

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/fragments/PopupAttributesListFragment.kt:
```diff
@@ -43,11 +50,69 @@ class PopupAttributesListFragment : Fragment() {
+    /**
+     * Shows error message to the use in a dialog.
+     */
+    private fun showAlertDialog(message: String) {
```
I wonder if the two "show dialog" methods could be combined into one. Ideally put them in a helper class so it can be re-used in other places.

puneet-pdx replied:
The two show dialogs cannot be combined into one as one has just the cancel button(it dismisses the dialog when user presses it) and the second one has both OK and Cancel. 

In the 1st case when the user presses OK we just dismiss the dialog. 
In the 2nd case when the user presses OK we need to perform some follow up operation when user presses OK. 

The solution to get some value back when user presses the button is not very straight forward,
https://stackoverflow.com/questions/10763986/android-confirmation-dialog-returning-true-or-false

gunt0001 replied:
Ok, lets leave it as is for now. We can refactor if we need this logic in other places.

gunt0001 reviewed (commented):
@puneet-pdx - see my comments

puneet-pdx committed:
WIP

puneet-pdx committed:
WIP

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/viewmodels/IdentifyResultViewModel.kt:
```diff
@@ -23,57 +23,47 @@ import com.esri.arcgisruntime.data.Feature
     val showPopupAttributeListEvent: LiveData<Event<Unit>> = _showPopupAttributeListEvent
 
     /**
      * Highlights the result popup in the GeoView.
```
Can the doc be expanded? This function doesn't seem to just highlight a feature

puneet-pdx replied:
added doc

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/viewmodels/IdentifyResultViewModel.kt:
```diff
@@ -23,57 +23,47 @@ import com.esri.arcgisruntime.data.Feature
+        _showIdentifiedPopupAttributesEvent.raiseEvent()
+        highlightFeatureInFeatureLayer(
+            identifyLayerResult.layerContent,
+            identifyLayerResult.popups[0].geoElement
```
need to check for empty popup collection?

puneet-pdx replied:
updated

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/viewmodels/IdentifyResultViewModel.kt:
```diff
@@ -23,57 +23,47 @@ import com.esri.arcgisruntime.data.Feature
-            )
-        }
+        _identifyLayerResult.value = identifyLayerResult
+        _showIdentifiedPopupAttributesEvent.raiseEvent()
```
Does this show the attribute list? Why would we want to do that here?

puneet-pdx replied:
It shows the identified popup in the bottomsheet. I have updated the event name and the doc for this method.

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/viewmodels/PopupViewModel.kt:
```diff
@@ -0,0 +1,163 @@
+     * Clear the popup
+     */
+    fun clearPopup() {
+        _popup.value = null
```
Should also clear the PopupManager?

puneet-pdx replied:
updated

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/popup/PopupView.kt:
```diff
@@ -112,42 +108,93 @@ class PopupView : FrameLayout {
+         * Toggles the view for popup field value from edittext to textview and vice-versa, given the
+         * edit mode of the popupView.
+         */
+        fun updateViewMode(popupField: PopupField) {
```
better name: `updateView`?

puneet-pdx replied:
updated

gunt0001 commented on app/src/main/res/layout/fragment_popup_attribute_list.xml:
```diff
@@ -18,26 +18,68 @@
                     android:layout_width="wrap_content"
                     android:layout_height="wrap_content"
-                    app:imageBitmap="@{identifyResultViewModel.identifiedPopup}"
+                    app:imageBitmap="@{popupViewModel.popup}"
```
might be worth to add a comment here that mentions the used BindingAdapter

puneet-pdx replied:
added

puneet-pdx replied:
As per our discussion I have not added doc for this as we cannot add doc in the the xml tags.

gunt0001 commented on app/src/main/res/layout/fragment_popup_attribute_list.xml:
```diff
@@ -18,26 +18,68 @@
         xmlns:app="http://schemas.android.com/apk/res-auto">
```
If we decide to rename `PopupAttributesListFragment ` to `PopupFragment`, we also might want to rename this to `fragment_popup.xml`

puneet-pdx replied:
updated

gunt0001 reviewed (commented):
@puneet-pdx - looks good. A few more comments.

puneet-pdx committed:
moved popupViewModel to IdentifyResultFragment

puneet-pdx committed:
added doc

puneet-pdx committed:
Refactor event names

puneet-pdx committed:
remove doc

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/viewmodels/IdentifyResultViewModel.kt:
```diff
@@ -19,38 +19,48 @@ package com.esri.arcgisruntime.opensourceapps.datacollection.viewmodels
-    private val _showIdentifiedPopupAttributeEvent = MutableLiveData<Event<Unit>>()
-    val showIdentifiedPopupAttributeEvent: LiveData<Event<Unit>> = _showIdentifiedPopupAttributeEvent
+    private val _showIdentifyResultFragmentEvent = MutableLiveData<Event<Unit>>()
+    val showIdentifyResultFragmentEvent: LiveData<Event<Unit>> = _showIdentifyResultFragmentEvent
```
Avoid the term "Fragment" in a view model. Better: "showIdentifyResultEvent"

puneet-pdx replied:
updated

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/viewmodels/IdentifyResultViewModel.kt:
```diff
@@ -19,38 +19,48 @@ package com.esri.arcgisruntime.opensourceapps.datacollection.viewmodels
-    private val _identifiedPopup = MutableLiveData<Popup>()
-    val identifiedPopup: LiveData<Popup> = _identifiedPopup
+    private val _showPopupFragmentEvent = MutableLiveData<Event<Unit>>()
+    val showPopupFragmentEvent: LiveData<Event<Unit>> = _showPopupFragmentEvent
```
Avoid the term "Fragment" in a view model. Better: "showPopupEvent"

puneet-pdx replied:
updated

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/viewmodels/IdentifyResultViewModel.kt:
```diff
@@ -19,38 +19,48 @@ package com.esri.arcgisruntime.opensourceapps.datacollection.viewmodels
+     * <ul>
+     * <li>Sets the identifyLayerResult instance on the identifyResultViewModel
+     * <li>Sets the first popup returned by identifyLayerResult on PopupViewModel
+     * <li>Raises the event to show the popup in IdentifyResultFragment in the BottomSheet
```
I would only mention: "Raises an event to show the identify result"

How this identify result is actually presented, i.e. in a Fragment in a bottom sheet, is not the concern of the view model.

puneet-pdx replied:
updated

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/viewmodels/IdentifyResultViewModel.kt:
```diff
@@ -71,16 +81,23 @@ class IdentifyResultViewModel : ViewModel() {
+     * Kicks off IdentifyResultFragment
+     */
+    fun showIdentifiedPopupAttributes() {
+        _showIdentifyResultFragmentEvent.raiseEvent()
```
Shouldn't this raise the `showPopupFragmentEvent` event?

If so, I would also rename the method to `showPopup()`. The doc should not mention anything about a Fragment.

puneet-pdx replied:
updated

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/fragments/DataCollectionFragment.kt:
```diff
@@ -170,6 +181,7 @@ class DataCollectionFragment : Fragment() {
             when (destination.id) {
                 R.id.identifyResultFragment -> {
                     toolbar.title = dataCollectionViewModel.portalItemTitle.value
+                    popupViewModel.cancelEditing()
```
Were you going to remove this in a subsequent PR?

puneet-pdx replied:
We will need to handle the canceling editing of the popup when the user presses the back button while editing the popup. 
I intend to address this in the upcoming PR where I am cleaning up the UI and polishing some patterns.

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/fragments/DataCollectionFragment.kt:
```diff
@@ -242,7 +255,8 @@ class DataCollectionFragment : Fragment() {
         mapViewModel.identifiableLayer?.clearSelection()
-        identifyResultViewModel.resetIdentifyResult()
+        identifyResultViewModel.resetIdentifyLayerResult()
+        popupViewModel.clearPopup()
```
This call can be done in `identifyResultViewModel.resetIdentifyLayerResult()`

puneet-pdx replied:
updated

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/fragments/PopupFragment.kt:
```diff
@@ -0,0 +1,115 @@
+            false
+        )
+
+        val identifyResultViewModel: IdentifyResultViewModel by activityViewModels {
```
This doesn't seem right. A PopupFragment shouldn't have a dependency to the IdentifyResultViewModel. It should only be aware of the PopupViewModel.

puneet-pdx replied:
Removed. I was able to fix the issue which was related to the app crashing when the user will tap on the bottomsheet.

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/fragments/PopupFragment.kt:
```diff
@@ -0,0 +1,115 @@
+import kotlinx.android.synthetic.main.fragment_popup.*
+
+/**
+ * Responsible for displaying PopupAttribute list.
```
Better: "Responsible for displaying a Popup."

puneet-pdx replied:
updated

gunt0001 commented on app/src/main/res/layout/fragment_popup.xml:
```diff
@@ -0,0 +1,184 @@
+                android:id="@+id/controlLayout"
+                android:layout_width="match_parent"
+                android:layout_height="0dp"
+                android:onClick="@{() -> identifyResultViewModel.showIdentifiedPopupAttributes()}"
```
As mentioned above, this dependency doesn't seem right.

puneet-pdx replied:
Removed. I was able to fix the issue which was related to the app crashing when the user will tap on the bottomsheet.

When the user was tapping on the bottomsheet the tap event was not getting captured by the bottomsheet or the views that are in the bottomsheet, but was leaking to the underlying MapView, causing the view to load again and causing the crash. 
I am capturing the onClick in the bottomsheet now. So no need for this workflow for fixing the issue.

gunt0001 reviewed (commented):
@puneet-pdx - see my comments

puneet-pdx committed:
WIP

puneet-pdx committed:
addressed code review comments.

puneet-pdx committed:
update w600dp fragment_data_collection

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/viewmodels/PopupViewModel.kt:
```diff
@@ -0,0 +1,164 @@
+    /**
+     * Enables/disables edit mode on the PopupView
+     */
+    fun setEditMode(isEnable: Boolean) {
```
isEnabled

puneet-pdx replied:
updated

gunt0001 commented on app/src/main/java/com/esri/arcgisruntime/opensourceapps/datacollection/views/fragments/DataCollectionFragment.kt:
```diff
@@ -136,14 +142,19 @@ class DataCollectionFragment : Fragment() {
 
-        identifyResultViewModel.showPopupAttributeListEvent.observeEvent(viewLifecycleOwner) {
+        identifyResultViewModel.showPopupEvent.observeEvent(viewLifecycleOwner) {
             bottomSheetNavController.navigate(R.id.action_identifyResultFragment_to_popupAttributeListFragment)
```
I think the action should be renamed to `action_identifyResultFragment_to_popupFragment`

puneet-pdx replied:
updated

gunt0001 reviewed (approved):
@puneet-pdx - looks good, two minor comments.

puneet-pdx committed:
Refactor isEnable and PopupAttributeListFragment in bottomsheet navigation
